src/frontend/Desugar.ml

```diff
@@ -0,0 +1,97 @@
+open Core_kernel
+open Ast
+
+let is_multi_index = function
+  | Single {Ast.emeta= {Ast.type_= UArray _; _}; _}
+   |Downfrom _ | Upfrom _ | Between _ | All ->
+      true
+  | Single _ -> false
+
+let remove_trailing_alls_expr = function
+  | Indexed (obj, indices) ->
+      let rec remove_trailing_alls indices =
+        match List.rev indices with
+        | All :: tl -> remove_trailing_alls (List.rev tl)
+        | _ -> indices
+      in
+      Indexed (obj, remove_trailing_alls indices)
+  | e -> e
+
+(* TODO: same thing for Assignments *)
+let rec desugar_index_expr = function
+  | Indexed
+      ( { expr=
+            Indexed (obj, inner_indices)
+            (* , Single ({emeta= {type_= UArray UInt; _} as emeta; _} as multi)
+               *   :: inner_tl ) *)
+        ; emeta }
+      , (Single ({emeta= {type_= UInt; _}; _} as single_e) as single)
+        :: outer_tl )
+    when List.exists ~f:is_multi_index inner_indices -> (
+    match List.split_while ~f:(Fn.non is_multi_index) inner_indices with
+    | inner_singles, Single first_multi :: inner_tl ->
+        (* foo [arr1, ..., arrN] [i1, ..., iN] ->
+         foo [arr1[i1]] [arr[i2]] ... [arrN[iN]] *)
+        desugar_index_expr
+          (Indexed
+             ( { expr=
+                   Indexed
+                     ( obj
+                     , inner_singles
+                       @ [ Single
+                             { expr= Indexed (first_multi, [single])
+                             ; emeta= {emeta with type_= UInt} } ]
+                       @ inner_tl )
+               ; emeta }
+             , outer_tl ))
+    | inner_singles, Downfrom _ :: inner_tl | inner_singles, All :: inner_tl ->
+        (* v[:x][i] -> v[i] *)
+        (* v[:][i] -> v[i] *)
+        (* XXX generate check *)
+        desugar_index_expr
+          (Indexed
+             ( {expr= Indexed (obj, inner_singles @ [single] @ inner_tl); emeta}
+             , outer_tl ))
+    | inner_singles, Between (bot, _) :: inner_tl
+     |inner_singles, Upfrom bot :: inner_tl ->
+        (* v[x:y][z] -> v[x+z-1] *)
+        (* XXX generate check *)
+        desugar_index_expr
+          (Indexed
+             ( { expr=
+                   Indexed
+                     ( obj
+                     , inner_singles
+                       @ [ Single
+                             Frontend_utils.(
+                               binop (binop bot Plus single_e) Minus
+                                 (loop_bottom bot.emeta)) ]
+                       @ inner_tl )
+               ; emeta }
+             , outer_tl ))
+    | inner_singles, multis ->
```

Comment:
I think this would only match on multis = [], right? Don't you want a base case there?

Reply:
You are right that it will only match on `[]` now (I was putting in the multi stuff one by one before, oops). When you say base case do you just mean replacing `multis` with `[]` to be clear?

Reply:
Should it throw though? I would expect the behaviour on [] to be either doing nothing or just getting rid of the redundant indexing. That is
`a[] -> a`.

Reply:
the guard on the outer match is ` when List.exists ~f:is_multi_index inner_indices`, so it only triggers if the 2nd list isn't empty. It's just a vestigial throw because we can't let OCaml know that we know that list isn't empty.